web/src/components/predictions/predictions-game.js

```diff
@@ -1,21 +1,289 @@
-import React, { useState } from "react";
-import IntervalChoice from "./questions/interval-choice"
-import MultipleCategoryChoice from "./questions/multiple-category-choice"
+/** @jsx jsx */
+import React, {useEffect, useState} from 'react';
+import { Card, jsx, Text, Input, Label, Grid, Button, Box } from "theme-ui";
+import firebase from "gatsby-plugin-firebase";
+import { useList, useObject } from "react-firebase-hooks/database";
+import Spacer from "../../components/core/spacer";
+import IntervalChoice from "./questions/interval-choice";
+import MultipleCategoryChoice from "./questions/multiple-category-choice";
+import Leaderboard from "./leaderboard";
+import Login from "../users/login";
 
-function PredictionsGame(props) {
-  const lower = props.lower;
-  const upper = props.upper;
-  const [values, setValues] = useState([lower, upper]);
+const PredictionsGame = ({user}) => {
+  const [snapshot, loading, error] = useObject(
+    firebase.database().ref("predictions_users/" + user.uid)
+  );
+  const [questions, questionsLoading, questionsError] = useList(
+    firebase.database().ref("predictions/questions")
+  );
+  const [name, nameLoading, nameError] = useObject(
+      firebase.database().ref("predictions/leaderboard/" + user.uid)
+  );
+
+  // state hook for display name change
+  const [displayName, setDisplayName] = useState(user.displayName);
+  useEffect(() => {
+    setDisplayName(nameLoading ? displayName : name.child("nickname").val())
+  }, [nameLoading]);
+
+  const [borderColor, setBorderColor] = useState();
+  const [borderWidth, setBorderWidth] = useState(1);
+
+  // add user to firebase if doesn't exist
+  if (snapshot && !snapshot.exists()) {
+    if (!questionsLoading) {
+      let initial = {};
+      let initialScore = {};
+      initial["score"] = {};
+      questions.forEach((question) => (initial["score"][question.key] = 0));
+      initialScore["nickname"] = user.displayName;
+      initialScore["score"] = 0;
+      firebase
+        .database()
+        .ref("predictions_users/" + user.uid)
+        .set(initial);
+      firebase
+          .database()
+          .ref("predictions/leaderboard/" + user.uid)
+          .update(initialScore);
+    }
+  }
+
+  // answer is the index of the choice for MC questions, or the correct number for range questions
+  function calculateScore(isMC, prediction, answer, qid, uid, range) {
```

Comment:
Since this function is basically repeated in update-score.js, can we create a utils.js file to store these shared functions?